File: 907-koko-eating-bananas/koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # time: O(n * log(k)), k = max(piles), n = len(piles)
        low, high = 1, max(piles)
        res = high
        while low <= high:
            mid = (low + high) // 2

            time = 0
            for p in piles:
                time += math.ceil(p / mid)
            
            if time <= h:
                high = mid - 1
                res = mid
            else:
                low = mid + 1
        
        return res

File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        long l = 1;
        long r = Arrays.stream(piles).max().getAsInt();
        long res = r;
        while (l <= r) {
            long mid = (l + r) / 2;
            long hours = 0;
            for (int pile : piles) {
                hours += (int) Math.ceil((double) pile / mid);
            }

            if (hours <= h) {
                res = Math.min(res, mid);
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        }
        return (int) res;
    }
}